src/data/sqlite_converter.py

```diff
@@ -0,0 +1,389 @@
+    retro_big   = pd.DataFrame()
+    file_counter = 0
+    output_count = 0
+    gcd_count = 0
+    for u in range(len(input_files)):
+        gcd_dict, calibration = load_geospatial_data(gcd_files[u])
+        i3_file = dataio.I3File(input_files[u], "r")
+
+        while i3_file.more():
+            try:
+                frame = i3_file.pop_physics()
+            except:
+                frame = False
+            if frame:
+                truths, features, retros = process_frame(frame, mode, pulsemap, gcd_dict, calibration,input_files[u])
+                truth    = apply_event_no(truths, event_no_list, event_counter)
+                truth_big   = truth_big.append(truth, ignore_index = True, sort = True)
+                if len(retros)>0:
+                    retro   = apply_event_no(retros, event_no_list, event_counter)
+                    retro_big   = retro_big.append(retro, ignore_index = True, sort = True)
+                is_empty = isempty(features) 
+                if is_empty == False:
+                    features = apply_event_no(features, event_no_list, event_counter)
+                    feature_big= feature_big.append(features,ignore_index = True, sort = True)
+                event_counter += 1
+                if len(truth_big) >= max_dict_size:
+                    save_to_sql(feature_big, truth_big, retro_big, id, output_count, db_name,outdir, pulsemap)
+
+                    feature_big = pd.DataFrame()
+                    truth_big   = pd.DataFrame()
+                    retro_big   = pd.DataFrame()
+                    output_count +=1
+        file_counter +=1
+    if len(truth_big) > 0:
+        save_to_sql(feature_big, truth_big, retro_big, id, output_count, db_name, outdir, pulsemap)
+        feature_big = pd.DataFrame()
+        truth_big   = pd.DataFrame()
+        retro_big   = pd.DataFrame()
+        output_count +=1
+    return
+
+def save_to_sql(feature_big, truth_big, retro_big, id, output_count, db_name,outdir, pulsemap):
+    engine = sqlalchemy.create_engine('sqlite:///'+outdir + '/%s/tmp/worker-%s-%s.db'%(db_name,id,output_count))
+    truth_big.to_sql('truth',engine,index= False, if_exists = 'append')
+    if len(retro_big)> 0:
+        retro_big.to_sql('RetroReco',engine,index= False, if_exists = 'append')
+    feature_big.to_sql(pulsemap,engine,index= False, if_exists = 'append')
+    engine.dispose()
+    return
+
+class SQLiteDataConverter():
+    def __init__(self, paths, mode, pulsemap, gcd_rescue, outdir, db_name, workers,max_dictionary_size = 10000, verbose = 1):
+        self.paths          = paths
+        self.mode           = mode
+        self.pulsemap       = pulsemap
+        self.gcd_rescue     = gcd_rescue
+        self.outdir         = outdir
+        self.db_name        = db_name
+        self.verbose        = verbose
+        self.workers        = workers
+        self.max_dict_size  = max_dictionary_size 
+        self._extractor = I3Extractor()
+        self._processfiles()
```

Comment:
Right now this class does not use the `process_file` function of the base class (NB: I realise that this class does not, indeed, inherit from `DataConverter` right now, but eventually it will) and it does not seem to implement the `_save` method. This means that it deviates from the syntax "promised" by the base class. Therefore, I think that we should discuss whether the proposed syntax in `DataConverter` makes sense from an implementation standpoint; if so, we should rewrite the `SQLiteDataConverter` class to conform to this; if not, we should rewrite the base class to be a better fit to the actual implementation(s). Let me know what you think.

Reply:
Yes I'm breaking the rules already!

I think we should take a look at this class and from it decipher where we can generalize syntax for a practical general dataconverter class.

Rasmus